### PR TITLE
Support the lua scripts on kvrocks

### PR DESCRIPTION
**Pull request by git-hulk**
This PR was still working, mainly tasks contain:

- [x] Add a new flag to exclude the no script command, like `BPOP`
- [x] Also integrate the slowlog and perflog in Lua scripts
- [x] Check corner condition(still working but NOT affect the PR review)
- [x] Add test cases

### CODE REVIEW

**Comment by git-hulk** [comment]
No worry about conflicts, I'll fix thoses later.

**Comment by git-hulk** [comment]
Still 10,000 feet to achieve the whole functions of the Redis scripts, but at least the `eval` command works now. We would implement other scripting commands after this PR was merged.

**Comment by ShooterIT** [comment]
Amazing work !

Now only support `EVAL` command, don't support `EVALSHA`?

File: src/redis_cmd.cc [hunk 1]
```diff
@@ -183,7 +184,7 @@ class CommandPing : public Commander {
   }
 };
 
-class CommandSelect: public Commander {
+class CommandSelect : public Commander {
```
Comment (ShooterIT) [inline src/redis_cmd.cc]: I don't think it is a good idea to fix format style in this **big** PR

Comment (git-hulk) [reply src/redis_cmd.cc]: My bad, I should NOT format this file, would roll back those formats later.

**Comment by git-hulk** [comment]
> Amazing work !
> 
> Now only support `EVAL` command, don't support `EVALSHA`?

yes, would support it soon after this PR was merged.

File: src/server.h [hunk 2]
```diff
@@ -185,3 +196,8 @@ class Server {
   std::vector<WorkerThread *> worker_threads_;
   std::unique_ptr<ReplicationThread> replication_thread_;
 };
+
+extern Server *srv;
+
+Server *GetServer();
+
```
Comment (ShooterIT) [inline src/server.h]: unnecessary lines

File: kvrocks.conf [hunk 3]
```diff
@@ -483,4 +490,4 @@ rocksdb.stats_dump_period_sec 0
 # Default: no
 rocksdb.disable_auto_compactions no
 ################################ NAMESPACE #####################################
-# namespace.test change.me
+# namespace.test change.me
```
Comment (ShooterIT) [inline kvrocks.conf]: nee empty line

File: src/scripting.cc [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+    removeUnsupportedFunctions(lua);
+    loadFuncs(lua);
+    return lua;
+  }
+
+  void loadFuncs(lua_State *lua) {
+    lua_newtable(lua);
+
+    /* redis.call */
+    lua_pushstring(lua,  "call");
```
Comment (ShooterIT) [inline src/scripting.cc]: there are many places that two blankspaces in function

**Review by ShooterIT — APPROVED** [review]
LGTM, but don't read carefully!

We should add LUA memory usage in INFO command like redis, since we store lua in lua state now, it may eat much memory if there are too many scripts, right?

**Review by ShooterIT — DISMISSED** [review]
LGTM, but don't read carefully!

We should add LUA memory usage in INFO command like redis, since we store lua in lua state now, it may eat much memory if there are too many scripts, right?

**Comment by git-hulk** [comment]
> LGTM, but don't read carefully!
> 
> We should add LUA memory usage in INFO command like redis, since we store lua in lua state now, it may eat much memory if there are too many scripts, right?

yes, still have some necessary works to do in next PR, like:

* support `evalsha`,`scripting kill`
* Lua execution timeout

File: src/scripting.cc [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+
+    /* redis.call */
+    lua_pushstring(lua, "call");
+    lua_pushcfunction(lua, redisCallCommand);
+    lua_settable(lua,  -3);
+
+    /* redis.pcall */
+    lua_pushstring(lua, "pcall");
+    lua_pushcfunction(lua, redisPCallCommand);
+    lua_settable(lua,  -3);
```
Comment (ShooterIT) [inline src/scripting.cc:47]:
```suggestion
    lua_settable(lua, -3);
```

File: src/scripting.cc [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+    return lua;
+  }
+
+  void loadFuncs(lua_State *lua) {
+    lua_newtable(lua);
+
+    /* redis.call */
+    lua_pushstring(lua, "call");
+    lua_pushcfunction(lua, redisCallCommand);
+    lua_settable(lua,  -3);
```
Comment (ShooterIT) [inline src/scripting.cc]:
```suggestion
    lua_settable(lua, -3);
```

**Review by ShooterIT — DISMISSED** [review]

**Comment by ShooterIT** [comment]
currently, i wish to add LUA memory usage in INFO command, since lua eats too much memory but we have no idea to know, it is hard to find memory problem

**Review by ShooterIT — APPROVED** [review]

**Comment by git-hulk** [comment]
> currently, i wish to add LUA memory usage in INFO command, since lua eats too much memory but we have no idea to know, it is hard to find memory problem

aha, has already added, you can have a look at the last commit again
